don't leak build time path name to runtime for frozen python modules

shunting314 opened the pull request:
Summary:
Here is how we freeze a python module:
- we call python builtin compile method with the source code of the modules and the path. This method returns a python code object
- we call marshal.dumps to serialize the code object to bytes.

The code_object.co_filename actually matches the one passed in to the compile method. We can simply replace that with a marker
to avoid leak build time path to runtime.

This works on nested code objects as well:
```
#!/bin/env python3.8
import marshal

code_str = """
print("hello")

class MyCls:
    def __init__(self):
        pass
"""
co = compile(code_str, "<Generated by torch::deploy>", "exec")
cobytes = marshal.dumps(co)
import pdb; pdb.set_trace()
```

Checking `co`:
```
(Pdb) co.co_filename
'<Generated by torch::deploy>'
(Pdb) co.co_consts
('hello', <code object MyCls at 0x7f0e8670bbe0, file "<Generated by torch::deploy>", line 4>, 'MyCls', None)
(Pdb) co.co_consts[1].co_filename
'<Generated by torch::deploy>'
```

Test Plan:
Find the serialized frozenmodule for torch.nn.modules.linear module in the generated bytecode_x.c file. Put the content to /tmp/linear.bytecode

Run the testing script:
```
import marshal
co_bytes = bytes(eval("[{}]".format("".join(open('/tmp/linear.bytecode').readlines()).replace('\n', '').replace('\t', ''))))
co = marshal.loads(co_bytes)
print(co)

```

The output for the paste without the change:
```
<code object <module> at 0x7f39ca7f07c0, file "/data/users/shunting/fbsource/fbcode/buck-out/opt/gen/caffe2/gen_frozen_torchpython_src__srcs/torch/nn/modules/linear.py", line 1>
```

The output for the paste with the change:
```
<code object <module> at 0x7f05a765d710, file "<Generated by torch::deploy>", line 1>
````

Note that the file part is changed as expected.

Differential Revision: D31214555

suo reviewed (changes requested):
Change looks good to me, nice clean solution! Can we add a small unit test confirming the behavior?

suo commented on torch/csrc/deploy/interpreter/freeze.py:
```diff
@@ -225,7 +225,9 @@ def compile_file(self, path: Path, top_package_path: Path):
 
         with open(path, "r") as src_file:
-            co = compile(src_file.read(), path, "exec")
+            # instead of passing in the real build time path to 'compile', we pass in a marker instead
```
Also: Maybe expand this comment a little bit, explaining why we want to set a path marker and what problem it avoids.

suo commented on test/package/test_freeze.py:
```diff
@@ -0,0 +1,47 @@
+import textwrap
```
We'll need to make a few changes to the structure to support a pure Python test:
1. since this test covers `torch::deploy` and not `torch.package`, let's create a new `test/test_deploy.py` file and move that code there. You don't have to replicate the complicated structure fo the `test/package/` tree, since I don't think we'll have that many pure Python deploy tests—just a single general file called `test_deploy.py` is sufficient.
2. I'm not sure, but I don't think `torch/csrc/…` is technically guaranteed to be in PyTorch's Python install tree, since it's a not public library path (it's for C src). Let's move the `freeze.py` script to somewhere more conventional, like `torch/utils/_freeze.py` (note the forward underscore as freezing is not a user-facing module).

shunting314 replied:
Done. I've also changed the CMakeLists.txt under torch/csrc/deploy/interpreter. That cmake file will refer to the freeze script to create frozen python modules. Since we move the freeze script, we need change the commands in the cmake file accordingly.

suo commented on torch/csrc/deploy/interpreter/CMakeLists.txt:
```diff
@@ -74,7 +74,7 @@ add_custom_command(
    WORKING_DIRECTORY ${INTERPRETER_DIR}
    COMMAND mkdir -p ${FROZEN_DIR}
-   COMMAND ${PYTHON_BIN} freeze.py ${PYTHON_STDLIB_DIR} ${TYPING_PKG} ${PYTORCH_ROOT}/torch --oss --install_dir ${FROZEN_DIR} --verbose
+   COMMAND ${PYTHON_BIN} ../../../utils/_freeze.py ${PYTHON_STDLIB_DIR} ${TYPING_PKG} ${PYTORCH_ROOT}/torch --oss --install_dir ${FROZEN_DIR} --verbose
```
nit: `${PYTORCH_ROOT}/utils/_freeze.py` is more robust and readable.

shunting314 replied:
Indeed, too many dots on the path makes things easy to mess up. I got the idea. Looks like the correct path using PYTORCH_ROOT should be `${PYTORCH_ROOT}/torch/utils/_freeze.py` ?

suo replied:
yep!

suo reviewed (approved):
awesome! Lgtm assuming tests pass (you may need to change internal buck TARGETS as well)

shunting314 commented:
yep, all buck TARGETS depending on the freezing script I can find have been changed.

shunting314 committed:
don't leak build time path name to runtime for frozen python modules (#65715)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/65715

Here is how we freeze a python module:
- we call python builtin compile method with the source code of the modules and the path. This method returns a python code object
- we call marshal.dumps to serialize the code object to bytes.

The code_object.co_filename actually matches the one passed in to the compile method. We can simply replace that with a marker
to avoid leak build time path to runtime.

This works on nested code objects as well:
```
#!/bin/env python3.8
import marshal

code_str = """
print("hello")

class MyCls:
    def __init__(self):
        pass
"""
co = compile(code_str, "<Generated by torch::deploy>", "exec")
cobytes = marshal.dumps(co)
import pdb; pdb.set_trace()
```

Checking `co`:
```
(Pdb) co.co_filename
'<Generated by torch::deploy>'
(Pdb) co.co_consts
('hello', <code object MyCls at 0x7f0e8670bbe0, file "<Generated by torch::deploy>", line 4>, 'MyCls', None)
(Pdb) co.co_consts[1].co_filename
'<Generated by torch::deploy>'
```

Test Plan:
Find the serialized frozenmodule for torch.nn.modules.linear module in the generated bytecode_x.c file. Put the content to /tmp/linear.bytecode

Run the testing script:
```
import marshal
co_bytes = bytes(eval("[{}]".format("".join(open('/tmp/linear.bytecode').readlines()).replace('\n', '').replace('\t', ''))))
co = marshal.loads(co_bytes)
print(co)

```

The output for the paste without the change:
```
<code object <module> at 0x7f39ca7f07c0, file "/data/users/shunting/fbsource/fbcode/buck-out/opt/gen/caffe2/gen_frozen_torchpython_src__srcs/torch/nn/modules/linear.py", line 1>
```

The output for the paste with the change:
```
<code object <module> at 0x7f05a765d710, file "<Generated by torch::deploy>", line 1>
````

Note that the file part is changed as expected.

Reviewed By: suo

Differential Revision: D31214555

fbshipit-source-id: d70efa90cda644151089928895bf5bb2153c3e1d

shunting314 commented:
The failed check is unrelated to this PR.  The log (https://circleci.com/api/v1.1/project/github/pytorch/pytorch/16244115/output/107/0?file=true&allocation-id=61537bcb092ae40f7001046e-0-build%2F46508A2F ) says:

```
Sep 28 22:35:20   test_ind_worker_queue (__main__.TestIndividualWorkerQueue) ... ERROR (6.211s)

Too long with no output (exceeded 1h30m0s): context deadline exceeded
```

It seems some tests take too long and timeout.
